Trigger rebuilding of trento in obs

aleksei-burlakov opened the pull request:
(EDITED)
The first commit adds a GitHub action that calls obs services to make a package and then submits it to the OBS.
The second commit is to correctly rebase the PR over the main. (Lately the example/*.yaml cheks were removed)

stefanotorresi reviewed (changes requested):
@aleksei-burlakov to be honest, I'm having a lot of second thoughts about going through with the Ubuntu approach.

Copy-pasting the external code of the OBS source services is a major pitfall which I would strongly oppose to, and I just don't see the advantages, when we already have a [dedicated Docker image](https://github.com/arbulu89/continuous-delivery) that solved this problem for us.

By reusing the same approach we already have in place in the other projects, the entire workflow would be reduced just 3 steps:
- checkout
- initialize credentials
- execute OSC via already provided scripts

You wouldn't need to install any dependencies and we would be more in control of them, since they would be pinned and fixed, instead of potentially changing at every run of the CI job.

One last minor remark: in all the other projects, we usually put all this stuff under `packaging/suse` so that we allow for different packaging mechanism for other distros in the future.

stefanotorresi commented:
I also have one more question: how would we generate the OBS-compliant changelog?

aleksei-burlakov commented on obs/oscrc:
```diff
@@ -0,0 +1,6 @@
+apiurl = https://api.opensuse.org
+
+[https://api.opensuse.org]
+user=shapbot
```
I reuse the /scripts/init_osc_creds.sh in the docker image as described here https://confluence.suse.com/display/haslesforsap/Automatic+OBS+packages+updates+with+GitHub+and+GitHub+Actions

aleksei-burlakov commented on obs/specs/_service:
```diff
@@ -0,0 +1,27 @@
+        <param name="extract">web/frontend/package.json</param>
+        <param name="extract">web/frontend/package-lock.json</param>
+        <param name="exclude">web/frontend/package-lock.json</param>
+        <param name="versionformat">0.4.0+%ct.%h</param>
```
It's sedded as in https://confluence.suse.com/display/haslesforsap/Automatic+OBS+packages+updates+with+GitHub+and+GitHub+Actions  now

aleksei-burlakov replied:
> > > I also have one more question: how would we generate the OBS-compliant changelog?
> > 
> > 
> > By a shell script, I guess. I think we should address it in the next PR.
> 
> hey Aleksei, WRT this, maybe you can copy this:
> https://github.com/SUSE/sap_host_exporter/blob/master/.ci/gh_release_to_obs_changeset.py
> 
> Rest of the changes mostly LGTM, I will give it a closer look as soon as I can.

The problem with tockens is that the **$ obs token -o runservice** will fail, because the obs cannot download the vendors from outside.

aleksei-burlakov commented:
> I also have one more question: how would we generate the OBS-compliant changelog?

By a shell script, I guess. I think we should address it in the next PR.

aleksei-burlakov commented:
@here I have refactored the PR wrt to https://confluence.suse.com/display/haslesforsap/Automatic+OBS+packages+updates+with+GitHub+and+GitHub+Actions

stefanotorresi commented:
> > I also have one more question: how would we generate the OBS-compliant changelog?
> 
> By a shell script, I guess. I think we should address it in the next PR.

hey Aleksei, WRT this, maybe you can copy this:
https://github.com/SUSE/sap_host_exporter/blob/master/.ci/gh_release_to_obs_changeset.py

aleksei-burlakov commented:
> > > I also have one more question: how would we generate the OBS-compliant changelog?
> > 
> > 
> > By a shell script, I guess. I think we should address it in the next PR.
> 
> hey Aleksei, WRT this, maybe you can copy this:
> https://github.com/SUSE/sap_host_exporter/blob/master/.ci/gh_release_to_obs_changeset.py
> 
> Rest of the changes mostly LGTM, I will give it a closer look as soon as I can.

The problem with tockens is that the **$ obs token -o runservice** will fail, because the obs cannot download the vendors from outside.

stefanotorresi commented:
> The problem with tockens is that the **$ obs token -o runservice** will fail, because the obs cannot download the vendors from outside.

OBS tokens are not related to the changelog script I linked above: that uses GitHub tokens, which are available within GitHub actions jobs!

stefanotorresi commented on .github/workflows/submit-to-obs.yml:
```diff
@@ -0,0 +1,74 @@
+  TARGET_PROJECT: ${{ secrets.TARGET_PROJECT}}
+  FOLDER: packaging/suse
+jobs:
+  validation:
+    runs-on: ubuntu-18.04
+    steps:
+    - uses: actions/checkout@v2
+      with:
+        fetch-depth: 0
```
I guess this is not needed, right?

aleksei-burlakov replied:
Nope, it's there just in case we come up with some ideas about validation.

stefanotorresi commented on .github/workflows/submit-to-obs.yml:
```diff
@@ -0,0 +1,74 @@
+  delivery:
+    needs: validation
+    runs-on: ubuntu-18.04
+    if: ${{ github.event_name != 'pull_request' }}
```
this should actually only happen on the main branch:
```suggestion
    if: github.ref == 'refs/heads/main'
```

aleksei-burlakov replied:
:+1:

stefanotorresi commented on .github/workflows/submit-to-obs.yml:
```diff
@@ -0,0 +1,74 @@
+        fetch-depth: 0
+
+  delivery:
+    needs: validation
```
we don't really need any validation step

aleksei-burlakov replied:
Ok

stefanotorresi commented on .github/workflows/submit-to-obs.yml:
```diff
@@ -0,0 +1,74 @@
+        rpm --import ${OBS_UNSTABLE_KEY}
+        zypper ar ${OBS_UNSTABLE}
+        zypper ref
+        zypper in -y obs-service-node_modules go
```
Ideally we wouldn't need to do any of this and we would instead update the Docker image, but we can do this another time.
That said, we need to pin to a specific go version, otherwise this is going to blow up in our faces sooner or later.

aleksei-burlakov replied:
:+1:

stefanotorresi commented on .github/workflows/submit-to-obs.yml:
```diff
@@ -0,0 +1,74 @@
+        go version
+    - name: deliver package
+      run: |
+        echo "github.sha = ${{ github.sha }}"
+        echo "github.repository = ${{ github.repository }}"
```
I guess you put these there just for debugging?

aleksei-burlakov replied:
Yes, I will remove them.

stefanotorresi commented on .github/workflows/submit-to-obs.yml:
```diff
@@ -0,0 +1,74 @@
+      run: |
+        echo "github.sha = ${{ github.sha }}"
+        echo "github.repository = ${{ github.repository }}"
+        sed -i 's~%%VERSION%%~${{ github.sha }}~' $FOLDER/_service && \
```
this should not be `VERSION`, it should be `REVISION` instead
```suggestion
        sed -i 's~%%REVISION%%~${{ github.sha }}~' $FOLDER/_service && \
```

aleksei-burlakov replied:
Correct

stefanotorresi commented on .github/workflows/submit-to-obs.yml:
```diff
@@ -0,0 +1,74 @@
+        cp /root/.config/osc/oscrc $HOME/.config/osc
+    - name: submit package
+      run: |
+       sed -i 's~%%VERSION%%~${{ github.sha }}~' $FOLDER/_service && \
```
this should not be `VERSION`, it should be `REVISION` instead
```suggestion
        sed -i 's~%%REVISION%%~${{ github.sha }}~' $FOLDER/_service && \
```

aleksei-burlakov replied:
Correct

stefanotorresi commented on packaging/suse/_service:
```diff
@@ -0,0 +1,27 @@
+    <service name="tar_scm" mode="disabled">
+        <param name="url">https://github.com/%%REPOSITORY%%.git</param>
+        <param name="scm">git</param>
+        <param name="revision">%%VERSION%%</param>
```
this should not be `VERSION`, it should be `REVISION` instead
```suggestion
        <param name="revision">%%REVISION%%</param>
```

aleksei-burlakov replied:
Correct

stefanotorresi commented on packaging/suse/_service:
```diff
@@ -0,0 +1,27 @@
+        <param name="extract">web/frontend/package.json</param>
+        <param name="extract">web/frontend/package-lock.json</param>
+        <param name="exclude">web/frontend/package-lock.json</param>
+        <param name="versionformat">0.4.0+%ct.%h</param>
```
This is still hard-coded, and this is the actual `%%VERSION%%` parameter. 😉 
It should probably be sedded as well, like this: https://github.com/SUSE/sap_host_exporter/blob/aaa32755fedd8b687dfbc6373a95dc6224cf18f6/Makefile#L74

aleksei-burlakov replied:
Correct

stefanotorresi commented on packaging/suse/trento.spec:
```diff
@@ -0,0 +1,114 @@
+
+Name:           trento
+# Version will be processed via set_version source service
+Version:        0.4.0+1630322269.39b75d7
```
I think we can leave this to `0` to avoid confusion with the comment.

aleksei-burlakov replied:
Ок

stefanotorresi reviewed (changes requested):
Not quite there yet 🛠️

aleksei-burlakov commented:
> > The problem with tockens is that the **$ obs token -o runservice** will fail, because the obs cannot download the vendors from outside.
> 
> OBS tokens are not related to the changelog script I linked above: that uses GitHub tokens, which are available within GitHub actions jobs!

Ah, ok, I thought it was about that we wanted to make the whole thing by tokens.

aleksei-burlakov commented:
@stefanotorresi thank you for reviewing, it's all correct that you say about the version and revision. I have changed in another branch and forgot to bring it to the PR branch.

aleksei-burlakov commented:
@stefanotorresi I have refactored the PR wrt to the comments.

- `gh_release_to_obs_changeset.py` generated the `trento.changes`
- `VERSION` and `REVISION` are used correctly in the `_service`
- `Version: 0` in the trento.spec (it's defaulted to 0, and later set by the obs services)
- `echo ...` removed from the github action
- `valudation`-step is removed from the github action.

Besides, I have removed the yaml checks from the trento.spec in the second commit.

arbulu89 commented on packaging/suse/trento.spec:
```diff
@@ -0,0 +1,106 @@
+Version:        0
+Release:        0
+License:        Apache-2.0
+Summary:        An open cloud-native web console improving on the life of SAP Applications administrators .
```
It looks like there is an extra space here before the dot

aleksei-burlakov replied:
Good catch)

arbulu89 commented on packaging/suse/trento.spec:
```diff
@@ -0,0 +1,106 @@
+Group:          System/Monitoring
+URL:            https://github.com/trento-project/trento
+Source:         %{name}-%{version}.tar.gz
+Source1:        vendor.tar.gz
```
Where are these source files coming from?
Are they created during the CI process?
If they are pushed manually to the repo, I would at least put a comment in top of them

aleksei-burlakov replied:
They are created by the obs services, when calling `$ osc service dr` command. We run this command in the CI, but it could be run manually the same.

arbulu89 replied:
Nice!

arbulu89 commented on packaging/suse/trento.spec:
```diff
@@ -0,0 +1,106 @@
+%description
+An open cloud-native web console improving on the life of SAP Applications administrators.
+
+Trento is a city on the Adige River in Trentino-Alto Adige/Suedtirol in Italy. [...] It is one of the nation''s wealthiest and most prosperous cities, [...] often ranking highly among Italian cities for quality of life, standard of living, and business and job opportunities. (source)
```
Double quote here in `nation` word (`''`)

arbulu89 commented on packaging/suse/trento.spec:
```diff
@@ -0,0 +1,106 @@
+%defattr(-,root,root)
+%doc *.md
+%doc docs/*.md
+%if 0%{?suse_version} >= 1500
```
This `license` split will bring issues in the future. I would vote to only use the `license` tag` and remove the `doc` options, as this is only used in previous versions to SLE12SP3

arbulu89 commented on packaging/suse/trento.spec:
```diff
@@ -0,0 +1,106 @@
+%endif
+%{_bindir}/%{shortname}
+%{_unitdir}/trento-agent.service
+%{_sbindir}/rc%{name}
```
What is this line `rc%{name}` doing?

aleksei-burlakov replied:
It's a rudiment. Removed.

arbulu89 commented on packaging/suse/trento.spec:
```diff
@@ -0,0 +1,106 @@
+
+# Install the systemd unit
+install -D -m 0644 hack/trento-agent.service %{buildroot}%{_unitdir}/trento-agent.service
+sed -i "s+/srv/trento/trento agent start+trento agent start+g" %{buildroot}%{_unitdir}/trento-agent.service
```
I don't quite understand this line.
If we need to do such a thing, wouldn't be better to have the correct `trento-agent.service` after all?
It can be changed in the future though

aleksei-burlakov replied:
Correct. I will change the trento-agent.service as well. Besides, as discussed with @fabriziosestito I will move the `trento-agent.service` file from the `hack` folder to the root of the project. I think it's no bad to do it right in this PR.

aleksei-burlakov replied:
@fabriziosestito @arbulu89 the `trento-agent.service` needs another improvement. Since there are no more bench-common yaml checks, the trento-agent service will fail to start.

arbulu89 reviewed (commented):
@aleksei-burlakov I have not checked fully the CI file. I guess that you already worked on this with @stefanotorresi .
In any case, where are we submitting the package after uploading it to our devel repo? To tumbleweed factory?

I have put some minor comments in the spec file
